### ISSUE TITLE
Revise segfault on small changes in large package (v1.8.0-beta3 regression from v1.7)

### ISSUE THREAD

**Issue by oschulz** [JuliaLang/julia#44954]
With v1.8.0-beta3, I run into a segfault immediately with Revise on a (large) package (BAT.jl), even with very small changes. I don't really have an MWE, and the behavior is intermittent. I've used Revise very extensively with the package on Julia v1.7 without any trouble.

Steps to reproduce (starting with a completly empty `DEPOT_PATH`), all steps in a single Julia session (segfault doesn't always occur):

```
julia> versioninfo()
Julia Version 1.8.0-beta3
Commit 3e092a2521 (2022-03-29 15:42 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: 128 × AMD EPYC 7702P 64-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, znver2)
  Threads: 64 on 128 virtual cores

julia> DEPOT_PATH
1-element Vector{String}:
 "/tmp/.julia"

(@v1.8) pkg> status
Status `/tmp/.julia/environments/v1.8/Project.toml` (empty project)

(@v1.8) pkg> add Revise@3.3.3

julia> using Revise

(@v1.8) pkg> activate --temp

(jl_xxInzK) pkg> dev BAT

julia> cd(dirname(dirname(Base.find_package("BAT"))))

shell> git checkout revise-test-1
Branch 'revise-test-1' set up to track remote branch 'revise-test-1' from 'origin'.
Switched to a new branch 'revise-test-1'

(jl_TK0fkk) pkg> precompile

julia> using BAT

julia> bat_sample(BAT.example_posterior());

shell> git checkout revise-test-2
Branch 'revise-test-2' set up to track remote branch 'revise-test-2' from 'origin'.
Switched to a new branch 'revise-test-2'

julia> bat_sample(BAT.example_posterior());

signal (11): Segmentation fault
in expression starting at none:1
mtcache_hash_lookup at /buildworker/worker/package_linux64/build/src/typemap.c:292 [inlined]
jl_typemap_level_assoc_exact at /buildworker/worker/package_linux64/build/src/typemap.c:1025
jl_typemap_assoc_exact at /buildworker/worker/package_linux64/build/src/julia_internal.h:1324 [inlined]
jl_typemap_level_assoc_exact at /buildworker/worker/package_linux64/build/src/typemap.c:1027
jl_typemap_assoc_exact at /buildworker/worker/package_linux64/build/src/julia_internal.h:1324 [inlined]
jl_lookup_generic_ at /buildworker/worker/package_linux64/build/src/gf.c:2480 [inlined]
ijl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2536
```

This runs fine on Julia v1.7.2 (segfault never occurs).

More compact version of steps-to-reproduce, without switching REPL modes:

```
using Pkg
pkg"add Revise@3.3.3"
using Revise

pkg"activate --temp"
pkg"dev BAT"
cd(dirname(dirname(Base.find_package("BAT"))))

run(`git checkout revise-test-1`)

pkg"precompile"
using BAT
bat_sample(BAT.example_posterior());

run(`git checkout revise-test-2`)

bat_sample(BAT.example_posterior());
```

**Comment by gbaraldi** [comment]
might be similar to https://github.com/JuliaLang/julia/issues/44913

**Comment by vtjnash** [comment]
I can't say anything with the truncated backtrace. I attempted to run this on master a couple times, but it worked for me so far. Could you run with `--check-bounds=yes --bug-report=rr`?

**Comment by oschulz** [comment]
Here's a new version of the test script using two different BAT branches, this one seems to be more reproducible:


```
using Pkg
pkg"add Revise@3.3.3"
using Revise

pkg"activate --temp"
pkg"dev BAT"
cd(dirname(dirname(Base.find_package("BAT"))))

run(`git checkout revise-test-3`)

pkg"precompile"
using BAT

run(`git checkout revise-test-4`)

1+1
```

Running with `--check-bounds=yes --bug-report=rr` ... seems to take a while. :-)

**Comment by oschulz** [comment]
@vtjnash I tried running with `--check-bounds=yes --bug-report=rr` (after setting `sysctl -w kernel.perf_event_paranoid=1`) but it stalls after filling my `/tmp` directory to 100% (i.e. writing 100GB of data).

I can run with `--check-bounds=yes` and the updated script above reproduces the segfault, I get

```
...
julia> run(`git checkout revise-test-4`)

julia> 1+1

signal (11): Segmentation fault
in expression starting at none:1
mtcache_hash_lookup at /buildworker/worker/package_linux64/build/src/typemap.c:292 [inlined]
jl_typemap_level_assoc_exact at /buildworker/worker/package_linux64/build/src/typemap.c:1025
jl_typemap_assoc_exact at /buildworker/worker/package_linux64/build/src/julia_internal.h:1324 [inlined]
jl_typemap_level_assoc_exact at /buildworker/worker/package_linux64/build/src/typemap.c:1027
jl_typemap_assoc_exact at /buildworker/worker/package_linux64/build/src/julia_internal.h:1324 [inlined]
jl_lookup_generic_ at /buildworker/worker/package_linux64/build/src/gf.c:2480 [inlined]
ijl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2536
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1831 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:126
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:215
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:166 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:612
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:750
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:906
ijl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:965
eval at ./boot.jl:368 [inlined]
do_assignment! at /tmp/oschulz/julia-revisedebug/.julia/packages/JuliaInterpreter/7spiM/src/interpret.jl:354
step_expr! at /tmp/oschulz/julia-revisedebug/.julia/packages/JuliaInterpreter/7spiM/src/interpret.jl:470
step_expr! at /tmp/oschulz/julia-revisedebug/.julia/packages/JuliaInterpreter/7spiM/src/interpret.jl:594
finish! at /tmp/oschulz/julia-revisedebug/.julia/packages/JuliaInterpreter/7spiM/src/commands.jl:14
step_expr! at /tmp/oschulz/julia-revisedebug/.julia/packages/JuliaInterpreter/7spiM/src/interpret.jl:513
#methods_by_execution!#24 at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/lowered.jl:451
unknown function (ip: 0x7f863c14f61e)
unknown function (ip: 0x7f86bc12e349)
unknown function (ip: 0x7f86bc12e30d)
methods_by_execution!##kw at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/lowered.jl:239 [inlined]
#methods_by_execution!#20 at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/lowered.jl:217
methods_by_execution!##kw at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/lowered.jl:175 [inlined]
#eval_with_signatures#90 at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/packagedef.jl:464 [inlined]
eval_with_signatures##kw at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/packagedef.jl:462 [inlined]
#82 at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/packagedef.jl:319
unknown function (ip: 0x7f863c15a68f)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2358 [inlined]
ijl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2540
with_logstate at ./logging.jl:511
with_logger at ./logging.jl:623 [inlined]
#eval_rex#81 at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/packagedef.jl:312 [inlined]
eval_rex##kw at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/packagedef.jl:311 [inlined]
#revise#96 at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/packagedef.jl:777
revise at /tmp/oschulz/julia-revisedebug/.julia/packages/Revise/VskYC/src/packagedef.jl:735
unknown function (ip: 0x7f86bc14822f)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2358 [inlined]
ijl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2540
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1831 [inlined]
jl_f__call_latest at /buildworker/worker/package_linux64/build/src/builtins.c:769
#invokelatest#2 at ./essentials.jl:729 [inlined]
invokelatest at ./essentials.jl:727
unknown function (ip: 0x7f86bc08a362)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2358 [inlined]
ijl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2540
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1831 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:126
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:215
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:166 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:612
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:750
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:906
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:850
ijl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:965
eval at ./boot.jl:368 [inlined]
eval_user_input at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/REPL/src/REPL.jl:151
repl_backend_loop at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/REPL/src/REPL.jl:247
start_repl_backend at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/REPL/src/REPL.jl:232
#run_repl#47 at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/REPL/src/REPL.jl:369
run_repl at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.8/REPL/src/REPL.jl:356
jfptr_run_repl_63590.clone_1 at /opt/julia-1.8/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2358 [inlined]
ijl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2540
#964 at ./client.jl:419
jfptr_YY.964_53574.clone_1 at /opt/julia-1.8/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2358 [inlined]
ijl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2540
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1831 [inlined]
jl_f__call_latest at /buildworker/worker/package_linux64/build/src/builtins.c:769
#invokelatest#2 at ./essentials.jl:729 [inlined]
invokelatest at ./essentials.jl:727 [inlined]
run_main_repl at ./client.jl:404
exec_options at ./client.jl:318
_start at ./client.jl:522
jfptr__start_58493.clone_1 at /opt/julia-1.8/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2358 [inlined]
ijl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2540
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1831 [inlined]
true_main at /buildworker/worker/package_linux64/build/src/jlapi.c:567
jl_repl_entrypoint at /buildworker/worker/package_linux64/build/src/jlapi.c:711
main at julia-1.8 (unknown line)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x400808)
Allocations: 41597971 (Pool: 41564364; Big: 33607); GC: 42
Segmentation fault (core dumped)
```

**Comment by oschulz** [comment]
Running the test script above in one go makes the disk explode, but running it in two stages works.

Running with `julia`:

```
using Pkg
pkg"add Revise@3.3.3 BugReporting"

pkg"activate @revisedebug"
pkg"dev BAT"
cd(dirname(dirname(Base.find_package("BAT"))))

run(`git checkout revise-test-3`)

pkg"precompile"
using BAT
```

Then, running with `julia --check-bounds=yes --bug-report=rr`:

```
using Pkg
pkg"activate @revisedebug"
using Revise
using BAT

cd(dirname(dirname(Base.find_package("BAT"))))
run(`git checkout revise-test-4`)

1+1
```

Result:

```
signal (11): Segmentation fault
in expression starting at none:1
mtcache_hash_lookup at /buildworker/worker/package_linux64/build/src/typemap.c:292 [inlined]
jl_typemap_level_assoc_exact at /buildworker/worker/package_linux64/build/src/typemap.c:1025
jl_typemap_assoc_exact at /buildworker/worker/package_linux64/build/src/julia_internal.h:1324 [inlined]
[...]
To upload a trace, please authenticate, by visiting:
[...]
[ Info: Uploading Trace directory
Uploaded to https://s3.amazonaws.com/julialang-dumps/reports/[...]
┌ Info: Debugged process failed. Unless you see rr errors above, the trace likely completed.
│   exitcode = 0
└   termsignal = 11
Segmentation fault (core dumped)
```

@vtjnash can you work with that?

**Comment by cafaxo** [comment]
This is a shot in the dark, but have you tried downgrading JuliaInterpreter to [0.9.11](https://github.com/JuliaDebug/JuliaInterpreter.jl/releases/tag/v0.9.11)?

With 0.9.12, Revise also seems to be crashing for me. (I am on Julia master)
I will add a more detailed report on the issue I am an experiencing later.

**Comment by oschulz** [comment]
> This is a shot in the dark, but does this issue still occur if you downgrade JuliaInterpreter to 0.9.11?

Just tried, I get the same segfault as before.

**Comment by vtjnash** [comment]
you seem to be on an old version of Julia (missing https://github.com/JuliaLang/julia/pull/44855), but that should be unrelated to the segfault

**Comment by vtjnash** [comment]
That comment is for @cafaxo, who appears to be using julia-master to test this

**Comment by giordano** [comment]
Is this fixed by https://github.com/JuliaDebug/JuliaInterpreter.jl/pull/534?

**Comment by SebastianAment** [comment]
I was running into the same issue with Julia 1.8.0-beta3, but updating to Revise v3.3.3 with JuliaInterpreter 0.9.13 appears to have fixed it.

**Comment by vtjnash** [comment]
Jeff also fixed it in #44974 (2 related bugs)

**Comment by oschulz** [comment]
Thanks @vtjnash and @JeffBezanson !!
